server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/AbstractDocumentsFileManager.java

```diff
@@ -0,0 +1,232 @@
+    protected final int numMetrics;
+    protected final TrackingDirectoryWrapper tmpDirectory;
+    protected final SegmentWriteState state;
+    protected int docSizeInBytes = -1;
+
+    public AbstractDocumentsFileManager(
+        SegmentWriteState state,
+        StarTreeField starTreeField,
+        List<MetricAggregatorInfo> metricAggregatorInfos
+    ) {
+        this.starTreeField = starTreeField;
+        this.tmpDirectory = new TrackingDirectoryWrapper(state.directory);
+        this.metricAggregatorInfos = metricAggregatorInfos;
+        this.state = state;
+        numMetrics = metricAggregatorInfos.size();
+    }
+
+    private void setDocSizeInBytes(int numBytes) {
+        if (docSizeInBytes == -1) {
+            docSizeInBytes = numBytes;
+        }
+        assert docSizeInBytes == numBytes;
+    }
+
+    /**
+     * Write the star tree document to file associated with dimensions and metrics
+     */
+    protected int writeStarTreeDocument(StarTreeDocument starTreeDocument, IndexOutput output, boolean isAggregatedDoc) throws IOException {
+        int numBytes = writeDimensions(starTreeDocument, output);
+        numBytes += writeMetrics(starTreeDocument, output, isAggregatedDoc);
+        setDocSizeInBytes(numBytes);
+        return numBytes;
+    }
+
+    /**
+     * Write dimensions to file
+     */
+    protected int writeDimensions(StarTreeDocument starTreeDocument, IndexOutput output) throws IOException {
+        int numBytes = 0;
+        for (int i = 0; i < starTreeDocument.dimensions.length; i++) {
+            output.writeLong(starTreeDocument.dimensions[i] == null ? 0L : starTreeDocument.dimensions[i]);
+            numBytes += Long.BYTES;
+        }
+        numBytes += StarTreeDocumentBitSetUtil.writeBitSet(starTreeDocument.dimensions, output);
+        return numBytes;
+    }
+
+    /**
+     * Write star tree document metrics to file
+     */
+    protected int writeMetrics(StarTreeDocument starTreeDocument, IndexOutput output, boolean isAggregatedDoc) throws IOException {
+        int numBytes = 0;
+        for (int i = 0; i < starTreeDocument.metrics.length; i++) {
+            switch (metricAggregatorInfos.get(i).getValueAggregators().getAggregatedValueType()) {
+                case LONG:
+                    output.writeLong(starTreeDocument.metrics[i] == null ? 0L : (Long) starTreeDocument.metrics[i]);
+                    numBytes += Long.BYTES;
+                    break;
+                case DOUBLE:
+                    if (isAggregatedDoc) {
+                        long val = NumericUtils.doubleToSortableLong(
+                            starTreeDocument.metrics[i] == null ? 0.0 : (Double) starTreeDocument.metrics[i]
+                        );
+                        output.writeLong(val);
+                    } else {
+                        output.writeLong(starTreeDocument.metrics[i] == null ? 0L : (Long) starTreeDocument.metrics[i]);
```

Comment:
Nope, this is in case of segment documents -> we are writing long to file, we're just reading DVP and storing whatever is present into file which will be converted later in sort and reduce flow